### PR TITLE
tools/nodetool: implement additional commands, part 2/N

### PR DESCRIPTION
**Pull request by denesb**
The following new commands are implemented:
* stop
* compactionhistory

All are associated with tests. All tests (both old and new) pass with both the scylla-native and the cassandra nodetool implementation.

Refs: https://github.com/scylladb/scylladb/issues/15588

### CODE REVIEW

**Comment by avikivity** [comment]
Here's a plan to incrementally switch nodetool:

1. rename /usr/bin/nodetool in tools/java to /usr/bin/nodetool-java
2. simultaneously, introduce /usr/bin/nodetool in scylladb.git that execs nodetool-java
3. make that pass CI (you'll need 2 to include the submodule update)
4. afterwards, change nodetool to first try the native version

File: tools/scylla-nodetool.cc [hunk 1]
```diff
@@ -239,6 +239,25 @@ void statusgossip_operation(scylla_rest_client& client, const bpo::variables_map
...
+    if (!vm.count("compaction_type")) {
+        throw std::invalid_argument("missing required parameter: compaction_type");
+    }
+
+    static const std::vector<std::string_view> recognized_compaction_types{"COMPACTION", "CLEANUP", "SCRUB", "RESHAPE", "RESHARD", "UPGRADE"};
+
+    const auto compaction_type = vm["compaction_type"].as<sstring>();
+
+    if (std::ranges::find(recognized_compaction_types, compaction_type) == recognized_compaction_types.end()) {
+        throw std::invalid_argument(fmt::format("invalid compaction type: {}", compaction_type));
```
Comment (avikivity) [inline tools/scylla-nodetool.cc:386]: avoiding comment about std::ranges::contains

**Review by avikivity — APPROVED** [review]

**Comment by avikivity** [comment]
Oh, and we'll need to change test.py's cluster manager too

File: tools/scylla-nodetool.cc (hunk 1 above)
Comment (denesb) [reply tools/scylla-nodetool.cc:386]: I'm very much looking forward to `std::ranges::contains()`, I don't understand what took them so long to introduce it.

**Comment by denesb** [comment]
> Here's a plan to incrementally switch nodetool:
> 
>     1. rename /usr/bin/nodetool in tools/java to /usr/bin/nodetool-java
> 
>     2. simultaneously, introduce /usr/bin/nodetool in scylladb.git that execs nodetool-java
> 
>     3. make that pass CI (you'll need 2 to include the submodule update)
> 
>     4. afterwards, change nodetool to first try the native version

There is one catch. `nodetool help` is implemented by scylla-nodetool, but it will only list commands it already has. This can be easily circumvented by always forwarding `nodetool help` to `nodetool-java`, until all commands are implemented indeed.

The plan looks good to me. @tchaikov do you see any pitfalls?

File: tools/scylla-nodetool.cc (hunk 1 above)
Comment (avikivity) [reply tools/scylla-nodetool.cc:386]: They always keep something good for the next version to get us to upgrade.

**Comment by tchaikov** [comment]
this is revision of the suggestion proposed by Avi in the slack. my concern comes from the packaging and its workflow. i am quoting my comment in slack here in case some of us missed it:

> regarding the packaging, i think we need to figure out what is the right thing to do and what is the expected behavior first.
> 
> currently, nodetools is packaged in java/dist/debian/debian, as part of the scylla-tools package. and this package depends on $product-conf. this does make sense to me. as the client side tool should not depend on server, but it needs the access the server configurations for talking to the cluster -- we don't support any form of zero-conf or DNS SRV records yet.
> 
> but with the proposed change, ideally, we scylla-tools should depends on usr/bin/scylla which is in turn provided by the  scylla-server package. please note scylla-server package includes all the bits to run scylla, like iotune, various configuration files.
> 
> but a client side tool does not require them, it's like what we want is just a xterm, but it depends on the whole xorg server . i think probably the first step is to extract usr/bin/scylla out of scylla-server into a smaller package, probably we can name it scylla-server-core and corresponding scylla-server-dbg. scylla-server-core will depend on scylla-conf.so that the client-side tools can require this package. (please note, i am using $product and scylla interchangeably when it comes to package name.)
> 
> and then, let scylla-server and scylla-tools depend on scylla-server-core.

since we always redistribute the unified package. it should be fine if we move one executable from one package to another. or we add a dependency in between two different packages. but what if user is using the packaging system to install scylla and its clients? the goal is to make sure the existing users of  the rpm/deb have a smooth upgrade path.

**Comment by avikivity** [comment]
We have a choice between
 - losing the ability to install scylla-tools without installing scylla-server (if we make scylla-tools dependent on scylla-server)
 - losing the ability to run nodetool remotely

I believe the first is more important.

We could probably refine the packaging granularity to bring it back:

scylla-bin - just the binaries, without any scripts or systemd integration, but with the magic thunk scripts for the relocatable binaries
scylla-nodetool - fluff to run nodetool, depends on scylla-bin
scylla-tools - the java stuff, depends on scylla-nodetool and scylla-conf
scylla-server - the same as today, but with a dependency on scylla-bin

this way, one can install scylla-tools (and avoid a full server installation) or scylla-server, or both.

File: tools/scylla-nodetool.cc (hunk 1 above)
Comment (nyh) [reply tools/scylla-nodetool.cc:386]: Especially "contains" was always one version behind the "find" function ;-)

**Comment by nyh** [comment]
> Here's a plan to incrementally switch nodetool:
> 
>     1. rename /usr/bin/nodetool in tools/java to /usr/bin/nodetool-java
> 
>     2. simultaneously, introduce /usr/bin/nodetool in scylladb.git that execs nodetool-java
> 
>     3. make that pass CI (you'll need 2 to include the submodule update)
> 
>     4. afterwards, change nodetool to first try the native version

In what sense is this plan "incremental"? It sounds like you still need to implement all nodetool subcommands in the "native version" before you can do step 4 - and before you do step 4 nothing actually changes.

I don't really think this is a big problem, though - nodetool is not huge, it doesn't have any really complicated commands, and in a few weeks or months, it can all be done and thoroughly tested with the new testing framework. At that point, we can switch cold-turkey (and of course leave time for more QA, fallout from dtests, and so on).

**Comment by nyh** [comment]
> We could probably refine the packaging granularity to bring it back:
> 
> scylla-bin - just the binaries, without any scripts or systemd integration, but with the magic thunk scripts for the relocatable binaries scylla-nodetool - fluff to run nodetool, depends on scylla-bin scylla-tools - the java stuff, depends on scylla-nodetool and scylla-conf scylla-server - the same as today, but with a dependency on scylla-bin
> 
> this way, one can install scylla-tools (and avoid a full server installation) or scylla-server, or both.

Another option to consider is to add a **build-time** option to build a Scylla executable that doesn't actually have the full Scylla in it - **just** the nodetool tool. Yes, it will be quite big for just nodetool, but we can make an effort to optimize that build to only include the object files that it really needs (unfortunately this will include Seastar, but not most of the bulk of the Scylla source code).

Then we can have an RPM with just scylla-nodetool, without Scylla, and it should (hopefully) be significantly smaller than the full Scylla.

**Comment by avikivity** [comment]
> > We could probably refine the packaging granularity to bring it back:
> > scylla-bin - just the binaries, without any scripts or systemd integration, but with the magic thunk scripts for the relocatable binaries scylla-nodetool - fluff to run nodetool, depends on scylla-bin scylla-tools - the java stuff, depends on scylla-nodetool and scylla-conf scylla-server - the same as today, but with a dependency on scylla-bin
> > this way, one can install scylla-tools (and avoid a full server installation) or scylla-server, or both.
> 
> Another option to consider is to add a **build-time** option to build a Scylla executable that doesn't actually have the full Scylla in it - **just** the nodetool tool. Yes, it will be quite big for just nodetool, but we can make an effort to optimize that build to only include the object files that it really needs (unfortunately this will include Seastar, but not most of the bulk of the Scylla source code).
> 
> Then we can have an RPM with just scylla-nodetool, without Scylla, and it should (hopefully) be significantly smaller than the full Scylla.

I don't think we need to optimize for package size. It's nice to have smaller packages but not at the cost of even more complexity.

**Comment by avikivity** [comment]
scylla-server rpm is 53MB.

Getting systemd integration when you install tools is unexpected. It's supposed to be just client-side tools, not a server that will surprise you later.

**Comment by nyh** [comment]
> scylla-server rpm is 53MB.
> 
> Getting systemd integration when you install tools is unexpected. It's supposed to be just client-side tools, not a server that will surprise you later.

I'm not an expert in our packaging system, but **installing** scylla-server shouldn't mean running it. If the tools package is already 53 MB and not smaller, it might as well contain the entire Scylla and have just one package - as long as it doesn't enable the server by default. Maybe what I'm saying now is similar to the scylla-bin package you suggested - which contains the big binary for everything, but not the front-end scripts. But I think that also adding all the front-end scripts won't really be a problem - as long as they are not run by default.

**Comment by avikivity** [comment]
> > scylla-server rpm is 53MB.
> > Getting systemd integration when you install tools is unexpected. It's supposed to be just client-side tools, not a server that will surprise you later.
> 
> I'm not an expert in our packaging system, but **installing** scylla-server shouldn't mean running it. 

It does not. But it has footprint in systemd (systemctl --all will show it), and it will install a bunch more scripts and stuff that is not expected.

> If the tools package is already 53 MB and not smaller, it might as well contain the entire Scylla and have just one package - as long as it doesn't enable the server by default. Maybe what I'm saying now is similar to the scylla-bin package you suggested - which contains the big binary for everything, but not the front-end scripts. But I think that also adding all the front-end scripts won't really be a problem - as long as they are not run by default.

If the tools package contains the scylla binary, then the server package can't contain it.

That's why I suggested a new package, so it can be shared between the two.

**Comment by denesb** [comment]
Update:
* Fix timezone issues (hopefully)

**Pull request merged**